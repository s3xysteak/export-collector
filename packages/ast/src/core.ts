export * from './core/index'
export * from './core/parser'
