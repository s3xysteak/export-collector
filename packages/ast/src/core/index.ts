import { promises as fs } from 'node:fs'
import process from 'node:process'
import { extname, resolve } from 'pathe'
import type { ExpGeneratorDataOptions, ExpGeneratorOptions } from './types'

import { expCollector } from './parser'
import { getPkg } from './utils'
import { COMMENT, importsTemplate, resolversTemplate } from '~shared/template'

/**
 * Entry
 */
export async function expGenerator(path: string, options: Partial<ExpGeneratorOptions> = {}) {
  const {
    data,
    pkgContext: { isTs },
  } = await expGeneratorData(path, options)

  const {
    type,
    cwd,

    typescript = isTs,
    writeTo = `./src/${type}.${typescript ? 'ts' : 'js'}`,
  } = resolveOptions(options)

  return await fs.writeFile(
    resolve(cwd, extname(writeTo) ? writeTo : `${writeTo}.${typescript ? 'ts' : 'js'}`),
    data,
  )
}

export async function expGeneratorData(path: string, options: Partial<ExpGeneratorDataOptions> = {}) {
  const { raw: pkg, isTs } = await getPkg()

  const {
    cwd,
    include,
    exclude,
    rename,
    exportDefault,
    alias,
    type,

    typescript = isTs,
    pkgName = pkg.name,
  } = resolveOptions(options)

  exclude.push(rename)

  const expList = await expCollector(path, { cwd, alias })

  const exportList = [...expList, ...include].filter(i => !exclude.includes(i)).sort()

  const val = type === 'imports'
    ? importsTemplate(exportList, pkgName, rename, exportDefault)[typescript ? 'ts' : 'js']
    : resolversTemplate(exportList, pkgName, rename, exportDefault)[typescript ? 'ts' : 'js']

  const data = `
${COMMENT}

${val.trim()}

${COMMENT}
`.trim()

  return {
    data: `${data}\n`,
    pkgContext: {
      pkg,
      isTs,
    },
  }
}

function resolveOptions(options: Partial<ExpGeneratorOptions> = {}) {
  const {
    type = 'imports',
    cwd = process.cwd(),
    include = [],
    exclude = [],
    rename = 'autoImport',
    exportDefault = true,
    alias,
    ...rest
  } = options

  return {
    type,
    cwd,
    include,
    exclude,
    rename,
    exportDefault,
    alias,
    ...rest,
  }
}
