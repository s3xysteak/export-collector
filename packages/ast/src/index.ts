export * from './core/index'
export * from './core/parser'

export { default } from './core/unplugin'
