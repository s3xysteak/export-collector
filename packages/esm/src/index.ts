export * from './core/parser'

export { default } from './core/unplugin'
