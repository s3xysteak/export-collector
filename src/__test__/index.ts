export * from './core/func1'
export * from './core/func2'
