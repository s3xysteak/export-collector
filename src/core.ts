export * from './core/parser'
export * from './core/index'
