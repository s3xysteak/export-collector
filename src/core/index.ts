import process from 'node:process'
import { promises as fs } from 'node:fs'
import { isUndefined } from '@s3xysteak/utils'
import { extname, resolve } from 'pathe'
import { expCollector } from './parser'
import { findPath, getPkg } from './utils'

import { COMMENT, generateTemplate, importsTemplate, resolversTemplate } from './template'

interface ExpGeneratorDataOptions {
  /**
   * The base path of the project.
   * @default process.cwd()
   */
  base: string

  /**
   * The package name of the project. The default value is the name of the `package.json`.
   * @default pkg.name
   */
  pkgName: string

  /**
   * The list of exports to include.
   * @default []
   */
  include: string[]

  /**
   * The list of exports to exclude.
   * @default []
   */
  exclude: string[]

  /**
   * The name of the export function.
   * @default 'autoImport'
   */
  rename: string

  /**
   * Whether to use TypeScript. The default value will be auto detected by `package.json`.
   * @default isTs
   */
  typescript: boolean

  /**
   * Whether to export the function as default.
   * @default false
   */
  exportDefault: boolean

  alias: Record<string, string>

  /**
   * unplugin-auto-import options
   * @default 'imports'
   */
  type: 'imports' | 'resolvers'
}

export interface ExpGeneratorOptions extends ExpGeneratorDataOptions {
  /**
   * The path to write the generated file. The default value is the path of the entry file.
   */
  writeTo: string
}
/**
 * Entry
 */
export async function expGenerator(path: string, options: Partial<ExpGeneratorOptions> = {}) {
  const {
    data,
    dataRaw,
    targetPath,
    pkgContext: { isTs },
  } = await expGeneratorData(path, options)

  const {
    base = process.cwd(),
  } = options

  return isUndefined(options?.writeTo)
    ? await fs.writeFile(targetPath, data)
    : await fs.writeFile(
      resolve(base, extname(options.writeTo) || `${options.writeTo}.${isTs ? 'ts' : 'js'}`),
      dataRaw,
    )
}

export async function expGeneratorData(path: string, options?: Partial<ExpGeneratorDataOptions>) {
  const { raw: pkg, isTs } = await getPkg()

  const {
    base = process.cwd(),
    pkgName = pkg.name,
    include = [],
    exclude = [],
    rename = 'autoImport',
    typescript = isTs,
    exportDefault = false,
    alias,
    type = 'imports',
  } = options ?? {}

  exclude.push(rename)

  const targetPath = await findPath(path, base)

  const expList = await expCollector(path, { base, alias })
  const content = await fs.readFile(targetPath, 'utf-8')

  const exportList = [...expList, ...include].filter(i => !exclude.includes(i)).sort()

  const valRaw = type === 'imports'
    ? importsTemplate(exportList, pkgName, rename, exportDefault)[typescript ? 'ts' : 'js']
    : resolversTemplate(exportList, pkgName, rename, exportDefault)[typescript ? 'ts' : 'js']

  const val = generateTemplate(content, valRaw).trim()
  const dataRaw = `
${COMMENT}

${valRaw.trim()}

${COMMENT}
`.trim()

  return {
    data: `${val}\n`,
    dataRaw: `${dataRaw}\n`,
    targetPath,
    pkgContext: {
      pkg,
      isTs,
    },
  }
}
