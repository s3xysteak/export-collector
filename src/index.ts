import { readFileSync } from 'node:fs'
import { parseSync } from '@swc/core'

export function getPath(rawPath: string) {
  // TODO
  return rawPath
}

export function parser(path: string): string[] {
  const result: string[] = []

  const module = parseSync(readFileSync(getPath(path), 'utf-8'), {
    syntax: 'typescript',
    target: 'es2020',
  })

  module.body.forEach((node) => {
    if (node.type === 'ExportAllDeclaration')
      parser(node.source.value)

    if (node.type === 'ExportNamedDeclaration') {
      node.specifiers.forEach((specifier) => {
        if (specifier.type === 'ExportSpecifier')
          result.push(specifier.orig.value)
        if (specifier.type === 'ExportNamespaceSpecifier')
          parser(specifier.name.value)
      })
    }

    if (node.type === 'ExportDeclaration') {
      node.declaration.type === 'VariableDeclaration' && node.declaration.declarations.forEach((declaration) => {
        declaration.id.type === 'Identifier' && result.push(declaration.id.value)
      })

      node.declaration.type === 'FunctionDeclaration' && result.push(node.declaration.identifier.value)
      node.declaration.type === 'ClassDeclaration' && result.push(node.declaration.identifier.value)
    }
  })

  return result
}
