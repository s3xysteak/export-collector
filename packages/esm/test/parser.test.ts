import { fileURLToPath } from 'node:url'
import { describe, expect, it } from 'vitest'
import { resolve } from 'pathe'
import { collect } from '../src'

const cwd = fileURLToPath(new URL('.', import.meta.url))

describe('esm parser', () => {
  it('collect', async () => {
    expect(await collect(resolve(cwd, './fixture/TS.ts'))).toMatchSnapshot()
    expect(await collect(resolve(cwd, './fixture/JS.js'))).toMatchSnapshot()
  })
})
