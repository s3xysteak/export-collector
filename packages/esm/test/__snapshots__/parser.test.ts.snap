// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`esm parser > collect 1`] = `
[
  "one",
  "shared",
  "two",
]
`;

exports[`esm parser > collect 2`] = `
[
  "one",
  "shared",
  "two",
]
`;
